File: internal/cli/format.go
package cli

import (
	"encoding/json"
	"fmt"
	"io"
	"strconv"
	"strings"
	"time"

	"hash/internal/engine"
)

// FormatOptions selects the output encoding and which optional columns are
// included. It replaces the boolean parameter list that used to make call
// sites unreadable.
type FormatOptions struct {
	Format       string
	ShowSize     bool
	ShowModified bool
	UTC          bool
}

// row is the flattened, per-algorithm view used by the text and TSV encoders.
// A failed file produces a single row with Err set.
type row struct {
	Algorithm string
	Digest    string
	Path      string
	Err       error
	Size      int64
	Modified  time.Time
}

// WriteResults renders results using options. The text and TSV encoders emit one
// row per algorithm; JSON emits one object per file.
func WriteResults(writer io.Writer, results []engine.FileResult, algorithms []string, options FormatOptions) error {
	switch strings.ToLower(options.Format) {
	case "text":
		return writeText(writer, rowsFromResults(results, algorithms), options)
	case "tsv":
		return writeTSV(writer, rowsFromResults(results, algorithms), options)
	case "json":
		return writeJSON(writer, results, options)
	default:
		return fmt.Errorf("%w %q", ErrUnsupportedFormat, options.Format)
	}
}

func rowsFromResults(results []engine.FileResult, algorithms []string) []row {
	rows := make([]row, 0, len(results))
	for _, item := range results {
		if item.Err != nil {
			rows = append(rows, row{Path: item.Request.Path, Err: item.Err})
			continue
		}
		for _, algorithm := range algorithms {
			rows = append(rows, row{
				Algorithm: algorithm,
				Digest:    item.Result.Digests[algorithm],
				Path:      item.Result.Path,
				Size:      item.Result.Size,
				Modified:  item.Result.Modified,
			})
		}
	}
	return rows
}

func writeText(writer io.Writer, rows []row, options FormatOptions) error {
	for _, item := range rows {
		if item.Err != nil {
			if _, err := fmt.Fprintf(writer, "ERROR\t%s\t%s\n", sanitize(item.Path), sanitize(item.Err.Error())); err != nil {
				return err
			}
			continue
		}
		if _, err := fmt.Fprintf(writer, "%s\t%s\t%s", item.Algorithm, item.Digest, sanitize(item.Path)); err != nil {
			return err
		}
		if options.ShowSize {
			if _, err := fmt.Fprintf(writer, "\tsize=%d", item.Size); err != nil {
				return err
			}
		}
		if options.ShowModified {
			if _, err := fmt.Fprintf(writer, "\tmodified=%s", formatTime(item.Modified, options.UTC)); err != nil {
				return err
			}
		}
		if _, err := fmt.Fprintln(writer); err != nil {
			return err
		}
	}
	return nil
}

func writeTSV(writer io.Writer, rows []row, options FormatOptions) error {
	columns := []string{"algorithm", "digest", "path", "error"}
	if options.ShowSize {
		columns = append(columns, "size")
	}
	if options.ShowModified {
		columns = append(columns, "modified")
	}
	if _, err := fmt.Fprintln(writer, strings.Join(columns, "\t")); err != nil {
		return err
	}
	for _, item := range rows {
		values := make([]string, 0, len(columns))
		if item.Err != nil {
			values = append(values, "ERROR", "", item.Path, item.Err.Error())
			if options.ShowSize {
				values = append(values, "-")
			}
			if options.ShowModified {
				values = append(values, "-")
			}
		} else {
			values = append(values, item.Algorithm, item.Digest, item.Path, "")
			if options.ShowSize {
				values = append(values, strconv.FormatInt(item.Size, 10))
			}
			if options.ShowModified {
				values = append(values, formatTime(item.Modified, options.UTC))
			}
		}
		for index := range values {
			values[index] = tsvValue(values[index])
		}
		if _, err := fmt.Fprintln(writer, strings.Join(values, "\t")); err != nil {
			return err
		}
	}
	return nil
}

func writeJSON(writer io.Writer, results []engine.FileResult, options FormatOptions) error {
	output := make([]jsonResult, 0, len(results))
	for _, item := range results {
		entry := jsonResult{Path: item.Request.Path, Digests: item.Result.Digests, Error: errorString(item.Err)}
		if options.ShowSize && item.Err == nil {
			entry.Size = &item.Result.Size
		}
		if options.ShowModified && item.Err == nil {
			modified := item.Result.Modified
			if options.UTC {
				modified = modified.UTC()
			}
			entry.Modified = &modified
		}
		output = append(output, entry)
	}
	encoder := json.NewEncoder(writer)
	encoder.SetIndent("", "  ")
	return encoder.Encode(output)
}

type jsonResult struct {
	Path     string            `json:"path"`
	Size     *int64            `json:"size,omitempty"`
	Modified *time.Time        `json:"modified,omitempty"`
	Digests  map[string]string `json:"digests,omitempty"`
	Error    string            `json:"error,omitempty"`
}

func errorString(err error) string {
	if err == nil {
		return ""
	}
	return err.Error()
}

func formatTime(value time.Time, utc bool) string {
	if utc {
		value = value.UTC()
	}
	return value.Format(time.RFC3339)
}

func sanitize(value string) string {
	return strconv.Quote(value)
}

func tsvValue(value string) string {
	trimmed := strings.TrimLeft(value, " \t\r\n")
	value = strings.NewReplacer("\\", "\\\\", "\t", "\\t", "\r", "\\r", "\n", "\\n").Replace(value)
	if trimmed != "" && strings.ContainsRune("=+-@", rune(trimmed[0])) {
		return "'" + value
	}
	return value
}

File: src/cli/format.cpp
#include "cli/format.h"

#include "common/quoting.h"

#include <algorithm>

namespace hash_core {
namespace {

QString formatTime(const QDateTime &value, bool utc) {
    const QDateTime adjusted = utc ? value.toUTC() : value;
    return adjusted.toString(Qt::ISODate);
}

QString jsonEscape(const QString &value) {
    QString escaped;
    escaped.reserve(value.size() + 2);
    for (const QChar character : value) {
        const char16_t code = character.unicode();
        switch (code) {
        case u'"':
            escaped += QStringLiteral("\\\"");
            break;
        case u'\\':
            escaped += QStringLiteral("\\\\");
            break;
        case u'\n':
            escaped += QStringLiteral("\\n");
            break;
        case u'\r':
            escaped += QStringLiteral("\\r");
            break;
        case u'\t':
            escaped += QStringLiteral("\\t");
            break;
        case u'<':
            escaped += QStringLiteral("\\u003c");
            break;
        case u'>':
            escaped += QStringLiteral("\\u003e");
            break;
        case u'&':
            escaped += QStringLiteral("\\u0026");
            break;
        case u'\u2028':
            escaped += QStringLiteral("\\u2028");
            break;
        case u'\u2029':
            escaped += QStringLiteral("\\u2029");
            break;
        default:
            if (code < 0x20) {
                escaped +=
                    QStringLiteral("\\u00") + QString::number(code, 16).rightJustified(4, u'0');
            } else {
                escaped += character;
            }
            break;
        }
    }
    return escaped;
}

// row is the flattened, per-algorithm view used by the text and TSV
// encoders. A failed file produces a single row with a non-empty error.
struct Row {
    QString algorithm;
    QString digest;
    QString path;
    QString error;
    qint64 size = 0;
    QDateTime modified;
};

QVector<Row> rowsFromResults(const QVector<FileResult> &results, const QStringList &algorithms) {
    QVector<Row> rows;
    for (const FileResult &item : results) {
        if (!item.error.isEmpty()) {
            Row row;
            row.path = item.request.path;
            row.error = item.error;
            rows.append(row);
            continue;
        }
        for (const QString &algorithm : algorithms) {
            Row row;
            row.algorithm = algorithm;
            row.digest = item.result.digestFor(algorithm);
            row.path = item.result.path;
            row.size = item.result.size;
            row.modified = item.result.modified;
            rows.append(row);
        }
    }
    return rows;
}

void writeText(QString *output, const QVector<Row> &rows, const FormatOptions &options) {
    for (const Row &item : rows) {
        if (!item.error.isEmpty()) {
            *output += QStringLiteral("ERROR\t") + goQuote(item.path) + u'\t' +
                       goQuote(item.error) + u'\n';
            continue;
        }
        *output += item.algorithm + u'\t' + item.digest + u'\t' + goQuote(item.path);
        if (options.showSize) {
            *output += QStringLiteral("\tsize=") + QString::number(item.size);
        }
        if (options.showModified) {
            *output += QStringLiteral("\tmodified=") + formatTime(item.modified, options.utc);
        }
        *output += u'\n';
    }
}

void writeTsv(QString *output, const QVector<Row> &rows, const FormatOptions &options) {
    QStringList columns{QStringLiteral("algorithm"), QStringLiteral("digest"),
                        QStringLiteral("path"), QStringLiteral("error")};
    if (options.showSize) {
        columns.append(QStringLiteral("size"));
    }
    if (options.showModified) {
        columns.append(QStringLiteral("modified"));
    }
    *output += columns.join(u'\t') + u'\n';
    for (const Row &item : rows) {
        QStringList values;
        if (!item.error.isEmpty()) {
            values << QStringLiteral("ERROR") << QString() << item.path << item.error;
            if (options.showSize) {
                values << QStringLiteral("-");
            }
            if (options.showModified) {
                values << QStringLiteral("-");
            }
        } else {
            values << item.algorithm << item.digest << item.path << QString();
            if (options.showSize) {
                values << QString::number(item.size);
            }
            if (options.showModified) {
                values << formatTime(item.modified, options.utc);
            }
        }
        for (QString &value : values) {
            value = tsvValue(value);
        }
        *output += values.join(u'\t') + u'\n';
    }
}

void writeJson(QString *output, const QVector<FileResult> &results, const FormatOptions &options) {
    if (results.isEmpty()) {
        *output += QStringLiteral("[]\n");
        return;
    }
    *output += QStringLiteral("[\n");
    for (int index = 0; index < results.size(); ++index) {
        const FileResult &item = results.at(index);
        const bool failed = !item.error.isEmpty();
        QStringList fields;
        fields.append(QStringLiteral("    \"path\": \"") + jsonEscape(item.request.path) + u'"');
        if (options.showSize && !failed) {
            fields.append(QStringLiteral("    \"size\": ") + QString::number(item.result.size));
        }
        if (options.showModified && !failed) {
            const QDateTime modified =
                options.utc ? item.result.modified.toUTC() : item.result.modified;
            fields.append(QStringLiteral("    \"modified\": \"") +
                          jsonEscape(formatTime(modified, false)) + u'"');
        }
        if (!failed) {
            // encoding/json sorts map keys, so the digests object is sorted
            // by algorithm name rather than request order.
            QStringList names;
            names.reserve(item.result.digests.size());
            for (const Digest &digest : item.result.digests) {
                names.append(digest.algorithm);
            }
            std::sort(names.begin(), names.end());
            QStringList entries;
            for (const QString &name : names) {
                entries.append(QStringLiteral("      \"") + jsonEscape(name) +
                               QStringLiteral("\": \"") + jsonEscape(item.result.digestFor(name)) +
                               u'"');
            }
            fields.append(QStringLiteral("    \"digests\": {\n") +
                          entries.join(QStringLiteral(",\n")) + QStringLiteral("\n    }"));
        }
        if (!item.error.isEmpty()) {
            fields.append(QStringLiteral("    \"error\": \"") + jsonEscape(item.error) + u'"');
        }
        *output += QStringLiteral("  {\n") + fields.join(QStringLiteral(",\n"));
        *output +=
            index + 1 < results.size() ? QStringLiteral("\n  },\n") : QStringLiteral("\n  }\n");
    }
    *output += QStringLiteral("]\n");
}

} // namespace

bool writeResults(const QVector<FileResult> &results, const QStringList &algorithms,
                  const FormatOptions &options, QString *output, QString *error) {
    const QString format = options.format.toLower();
    if (format == QStringLiteral("text")) {
        writeText(output, rowsFromResults(results, algorithms), options);
        return true;
    }
    if (format == QStringLiteral("tsv")) {
        writeTsv(output, rowsFromResults(results, algorithms), options);
        return true;
    }
    if (format == QStringLiteral("json")) {
        writeJson(output, results, options);
        return true;
    }
    *error = QStringLiteral("unsupported output format ") + goQuote(options.format);
    return false;
}

QString tsvValue(const QString &value) {
    int start = 0;
    while (start < value.size() && (value.at(start) == u' ' || value.at(start) == u'\t' ||
                                    value.at(start) == u'\r' || value.at(start) == u'\n')) {
        ++start;
    }
    QString replaced = value;
    replaced.replace(u'\\', QStringLiteral("\\\\"));
    replaced.replace(u'\t', QStringLiteral("\\t"));
    replaced.replace(u'\r', QStringLiteral("\\r"));
    replaced.replace(u'\n', QStringLiteral("\\n"));
    if (start < value.size()) {
        const char16_t first = value.at(start).unicode();
        if (first == u'=' || first == u'+' || first == u'-' || first == u'@') {
            return u'\'' + replaced;
        }
    }
    return replaced;
}

} // namespace hash_core
